packages/core/options.ts

```diff
@@ -102,31 +102,88 @@ export const fromSnakeCaseWithPrefix: (prefix: string) => KeyFormatSource = (
 	};
 };
 
+// Parameters can be passed in via environment variables, command line or
+// configuration file, and subsequently overwrite the default ones and each other.
+// The passed in values have to be set for externally provided parameters, e.g.
+// CLI parameters, before resolving the final options object.
+// Higher index means higher priority.
+export enum ParameterResolverIndex {
+	DefaultOptions = 1,
+	ConfigurationFile,
+	EnvironmentVariables,
+	CommandLineArguments,
+}
+type ParameterResolver = {
+	name: string;
+	transformKey: KeyFormatSource;
+	failOnUnknown: boolean;
+	object: object;
+};
+type ParameterResolvers = Record<ParameterResolverIndex, ParameterResolver>;
+const defaultResolvers: ParameterResolvers = {
+	[ParameterResolverIndex.DefaultOptions]: {
+		name: "Default options",
+		transformKey: fromCamelCase,
+		failOnUnknown: true,
+		object: defaultOptions,
+	},
+	[ParameterResolverIndex.ConfigurationFile]: {
+		name: "Configuration file",
+		transformKey: fromCamelCase,
+		failOnUnknown: true,
+		object: {},
+	},
+	[ParameterResolverIndex.EnvironmentVariables]: {
+		name: "Environment variables",
+		transformKey: fromSnakeCaseWithPrefix("JAZZER"),
+		failOnUnknown: false,
+		object: process.env as object,
+	},
+	[ParameterResolverIndex.CommandLineArguments]: {
+		name: "Command line arguments",
+		transformKey: fromSnakeCase,
+		failOnUnknown: true,
+		object: {},
+	},
+};
+
 /**
- * Builds a complete `Option` object based on default options, environment variables and
- * the partially given input options.
- * Keys in the given option object can be transformed by the `transformKey` function.
- * Environment variables need to be set in snake case with the prefix`JAZZER_`.
+ * Set the value object of a parameter resolver. Every resolver expects value
+ * object parameter names in a specific format, e.g. camel case or snake case,
+ * see the resolver definitions for details.
  */
-export function processOptions(
-	inputs: Partial<Options> = {},
-	transformKey: KeyFormatSource = fromCamelCase,
-	defaults: Options = defaultOptions,
-): Options {
+export function setParameterResolverValue(
+	index: ParameterResolverIndex,
+	inputs: Partial<Options>,
+) {
 	// Includes and excludes must be set together.
 	if (inputs && inputs.includes && !inputs.excludes) {
 		inputs.excludes = [];
 	} else if (inputs && inputs.excludes && !inputs.includes) {
 		inputs.includes = [];
 	}
+	defaultResolvers[index].object = inputs;
+}
+
+/**
+ * Build a complete `Option` object based on the parameter resolver chain.
+ * Add externally passed in values via the `setParameterResolverValue` function,
+ * before calling `buildOptions`.
+ */
+export function buildOptions(): Options {
+	const options = Object.keys(defaultResolvers)
+		.sort()
```

Comment:
Can you add a comment why the sort is necessary?

Reply:
Done.